rfcs/021 - Datastore Restructuring RFC.md

```diff
@@ -0,0 +1,82 @@
+
+It may also make sense for the storage layer to get the execution context, as well; then no transferrance of the key in question between OS threads would be necessary and since blocking mutations like that prevent other mutations on the same key from happening simultaneously, that might be a better use of CPU time on these OS threads, as well. But mixing Tokio async/await with real OS threads may be difficult/impossible, or may still require spawning a task with an `async move` block inside, which would nullify the benefits while complicating the storage layer, so we should be careful with this. Likely we don't do something like this in the first pass to make the layers cleaner and more easily extensible, and we come back to it later as an optimization pass when we're more convinced that little will change with datastore's API and the increased maintenance burden is worth it.
+
+![Datastore arch](./datastore_arch.jpg)
+
+The architecture and flow for the proposed datastore is diagrammed above. The routing layer (whether local mode or remote) is responsible for mutating the HandlerMemory for the given opcode and `opcode.rs` is simply a shell. It chooses the correct node to direct the work to and triggers its counterpart on the other node to request the relevant data from the storage layer, which then eventually returns it. The routing layer hands all of this to the execution layer on the same node to perform the requested work, which then returns it back to the routing layer on that node. The routing layer then stores the mutation, if any, back into the storage layer and then passes the result back to its counterpart on the original node, which then hands the results back to the original datastore opcode in the form of a new HandlerMemory object that replaces the old one (which may simply be the original HandlerMemory object in the case of `mutOnly` calls).
```

Comment:
Do you see an eventual bottleneck with this architecture with the routing layer?

Reply:
This part shouldn't be the bottleneck. There is a potential bottleneck with the way the datastore KV internals are structured, but the routing layer shouldn't be a problem.